prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -5467,67 +5496,226 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+
+        let usize_ty = self.encoder.env().tcx().mk_ty(ty::TyKind::Uint(ty::UintTy::Usize));
+        let idx_val_int = index.field(self.encoder.encode_value_field(usize_ty));
+
+        let lookup_pure_call = vir::Expr::func_app(
+            lookup_pure,
+            vec![
+                encoded_base_expr.clone(),
+                idx_val_int.clone()
+            ],
+            vec![
+                vir::LocalVar::new(
+                    String::from("self"),
+                    array_ty.clone(),
+                ),
+                vir::LocalVar::new(
+                    String::from("idx"),
+                    vir::Type::Int,
+                ),
+            ],
+            elem_value_ty,
+            vir::Position::default(),
+        );
+
+        stmts.push(vir::Stmt::Assert(
+            vir!{ [idx_val_int.clone()] < [ vir::Expr::from(array_len) ] },
+            vir::Position::default(),
+        ));
+        stmts.push(vir::Stmt::Assert(
+            vir::Expr::predicate_access_predicate(array_pred, encoded_base_expr, vir::PermAmount::Read),
+            vir::Position::default(),
+        ));
+
+        stmts.push(vir::Stmt::Inhale(vir!{ [ lookup_pure_call ] == [ lookup_res_val_field ] }));
+        Ok((lookup_res, stmts))
+    }
+
+    fn encode_array_lookup_mut(
+        &mut self,
+        base: PlaceEncoding<'tcx>,
+        index: vir::Expr,
+        array_ty: ty::Ty<'tcx>,
+        loan: Option<Borrow>,
+    ) -> EncodingResult<(vir::Expr, Vec<vir::Stmt>)> {
+        let (array_pred, array_ty, elem_pred, elem_ty, elem_value_ty, elem_ty_rs, array_len) =
+            self.encode_array_types(array_ty)?;
+
+        let res: vir::Expr = self.cfg_method.add_fresh_local_var(elem_ty.clone()).into();
+        let val_field = self.encoder.encode_value_field(elem_ty_rs);
+        let res_val_field = res.clone().field(val_field.clone());
+
+        let (encoded_base_expr, mut stmts) = self.postprocess_place_encoding(base, ArrayAccessKind::Mutable(None))?;
+
+        let usize_ty = self.encoder.env().tcx().mk_ty(ty::TyKind::Uint(ty::UintTy::Usize));
+        let idx_val_int = index.clone().field(self.encoder.encode_value_field(usize_ty));
+
+        // var res: Ref := havoc_ref()
+        stmts.extend(self.encode_havoc_and_allocation(&res));
+
+        // label before for old[label](..) stuff
+        let before_label = self.cfg_method.get_fresh_label_name();
+        stmts.push(vir::Stmt::label(&before_label));
+
+        // exhale preconditions
+        stmts.push(vir::Stmt::Assert(
+            vir!{ [idx_val_int.clone()] < [ vir::Expr::from(array_len) ] },
+            vir::Position::default(),
+        ));
+
+        // exhale Array$4$i32(self)
+        let array_access_pred = vir::Expr::predicate_access_predicate(
+            array_pred.clone(),
+            encoded_base_expr.clone(),
+            vir::PermAmount::Write,
+        );
+
+        stmts.push(vir!{ assert [array_access_pred.clone()] });
```

Comment:
It should be `exhale`, not `assert`.

Reply:
yes; this is the change mentioned in my lengthy comment to make it work a bit more, will change back of course.

Reply:
I see, now I understand the exhale/assert in the lengthy comment. Yes, it's probably because the magic wand is never applied.